bookmark-service/src/main/java/fr/loicmathieu/bookmarkit/BookmarkResource.java

```diff
@@ -58,34 +78,46 @@ public Bookmark get(@PathParam("id") Long id) {
     @Transactional
     @Operation(summary = "Create a bookmark")
     @Counted(name = "create.count")
-    @Timed(name="create.time")
-    public Response create(Bookmark bookmark){
-        bookmark.persist();
-        emitter.send(bookmark);
-        return Response.created(URI.create("/bookmarks/" + bookmark.id)).build();
+    @Timed(name = "create.time")
+    @Retry(maxRetries = 2)
+    public Response create(@Valid Bookmark bookmark) {
+        GeoIp geoIp = geoIpService.getIpInfos();
+        if (geoIp != null) {
+            bookmark.location = geoIp.city;
+            bookmark.persist();
+            this.emitter.send(bookmark);
+            return Response.status(Response.Status.CREATED).entity(bookmark).build();
+        } else {
+            return Response.status(Response.Status.INTERNAL_SERVER_ERROR).build();
+        }
     }
 
     @PUT
-    @Path("/{id}")
+    @Path("{id}")
     @Transactional
     @Operation(summary = "Update a bookmark")
     @Counted(name = "update.count")
-    @Timed(name="update.time")
-    public void update(Bookmark bookmark){
-        Bookmark existing = Bookmark.findById(bookmark.id);
-        existing.url = bookmark.url;
-        existing.description = bookmark.description;
-        existing.title = bookmark.title;
+    @Timed(name = "update.time")
+    public Response update(@Valid Bookmark bookmark, @PathParam("id") Long id, @Context UriInfo uriInfo) {
+        Bookmark entity = Bookmark.findById(id);
+        entity.description = bookmark.description;
+        entity.location = bookmark.location;
+        entity.title = bookmark.title;
+        entity.url = bookmark.url;
+        return Response.created(URI.create(uriInfo.getPath())).build();
     }
 
     @DELETE
-    @Path("/{id}")
+    @Path("{id}")
     @Transactional
     @Operation(summary = "Delete a bookmark")
     @Counted(name = "delete.count")
-    @Timed(name="delete.time")
-    public void delete(@PathParam("id")Long id){
-        Bookmark existing = Bookmark.findById(id);
-        existing.delete();
+    @Timed(name = "delete.time")
+    public Response delete(@PathParam("id") Long id) {
+        Bookmark bookmark = Bookmark.findById(id);
+        if (bookmark.isPersistent()) {
+            bookmark.delete();
+        }
```

Comment:
As we just retrieve the entity before, `isPersistent` never return false.

But as the findById method can return null this is the true check that we need to do (and throw a NotFoundException).Or we can just let the NPE happens, it's just a small dojo no need to have a fully covered code :)